File: go/015.go
package main

import "sort"

/**
给你一个整数数组 nums ，判断是否存在三元组之和为0且不重复。
思路：排序 + 双指针
-1,0,1,2,-1,-4   ->  -4,-1,-1,0,1,2
排序数组，然后遍历每个元素，为每个元素匹配条件sum == 0，使用双指针sum = s[i] + s[l] + s[r]
sum > 0 r--
sum < 0 l++
*重复元素跳过 写if时考虑要周全
*/
func threeSum(nums []int) [][]int {
	res, n := make([][]int, 0), len(nums)
	sort.Ints(nums)
	if n < 2 || nums[0] > 0 {
		return nil
	}
	for i := 0; i < n; i++ {
		l, r := i+1, n-1
		for i < n-1 && nums[i] == nums[i+1] {
			i++
		}
		for l < r {
			sum := nums[i] + nums[l] + nums[r]
			if sum < 0 {
				l++
			} else if sum > 0 {
				r--
			} else {
				res = append(res, []int{nums[i], nums[l], nums[r]})
				for l < r && nums[l] == nums[l+1] {
					l++
				}
				for l < r && nums[r] == nums[r-1] {
					r--
				}
				l++
				r--
			}
		}
	}
	return res
}

File: c++/015.cpp
class Solution {
/*
「双指针」：当我们需要枚举数组中的两个元素时，如果我们发现随着第一个元素下标的递增，第二个元素下标是递减的，那么就可以使用双指针的方法，将枚举的时间复杂度从 O(N^2) 减少至 O(N)
*/
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        int n = nums.size();
        sort(nums.begin(),nums.end());
        for(int k=0; k<n-2; k++){   //枚举至少保证最后有三个及以上的数
            if(nums[k]>0)      //因为已排好序，故后面的数的和必为正数
                break;
            if(k>0&&nums[k]==nums[k-1]) //过滤掉相同的数，避免结果重复
                continue;
            int i=k+1, j=n-1, t=-nums[k];
            while(i<j){     //双指针
                if(nums[i]+nums[j]==t){
                    res.push_back({nums[k],nums[i],nums[j]});
                    while(i<j&&nums[i]==nums[i+1])  //过滤重复的数，避免结果重复
                        i++;
                    while(i<j&&nums[j]==nums[j-1])
                        j--;
                i++;j--;        
                }else if(nums[i]+nums[j]<t){
                    while(i<j&&nums[i]==nums[i+1])
                            i++;
                    i++;
                }else{
                    while(i<j&&nums[j]==nums[j-1])
                            j--;
                    j--;
                }                
            }
        }
        return res;
    }
};